pkg/logger/logger.go

```diff
@@ -0,0 +1,269 @@
+	lc.rwMutex.RLock()
+	defer lc.rwMutex.RUnlock()
+	dump := make(map[LogOrigin]uint32, len(lc.data))
+	for k, v := range lc.data {
+		dump[k] = v
+	}
+
+	return dump
+}
+
+func newLogCounter() *LogCounter {
+	return &LogCounter{
+		rwMutex: sync.RWMutex{},
+		data:    map[LogOrigin]uint32{},
+	}
+}
+
+func getCallerInfo(skip int) (pkg, file string, line int) {
+	pc, file, line, ok := runtime.Caller(skip + 1)
+	if !ok {
+		panic("could not get runtime caller information")
+	}
+
+	funcName := runtime.FuncForPC(pc).Name()
+	lastSlash := strings.LastIndexByte(funcName, '/')
+	if lastSlash < 0 {
+		lastSlash = 0
+	} else {
+		lastSlash++
+	}
+	lastDot := strings.LastIndexByte(funcName, '.')
+	pkg = funcName[lastSlash:lastDot]
+	// check if it's from a receiver
+	if possibleLastDot := strings.LastIndexByte(pkg, '.'); possibleLastDot != -1 {
+		pkg = pkg[0:possibleLastDot]
+	}
+
+	return pkg, file, line
+}
+
+func (l *Logger) updateCounter(file string, line int) (new bool) {
+	return l.LogCount.update(LogOrigin{
+		File: file,
+		Line: line,
+	})
+}
+
+// Log functions
+
+// Debug
+func debug(skip int, l *Logger, msg string, fields ...Field) {
+	if l.cfg.Level > DebugLevel {
+		return
+	}
+
+	_, file, line := getCallerInfo(skip + 1)
+	if new := l.updateCounter(file, line); new {
+		l.l.Debug(msg, fields...)
+	}
+}
+
+func Debug(msg string, fields ...Field) {
+	debug(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) Debug(msg string, fields ...Field) {
+	debug(1, l, msg, fields...)
+}
+
+// Error
+func err(skip int, l *Logger, msg string, fields ...Field) {
+	_, file, line := getCallerInfo(skip + 1)
+	if new := l.updateCounter(file, line); new {
+		l.l.Error(msg, fields...)
+	}
+}
+
+func Error(msg string, fields ...Field) {
+	err(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) Error(msg string, fields ...Field) {
+	err(1, l, msg, fields...)
+}
+
+// Fatal
+func fatal(skip int, l *Logger, msg string, fields ...Field) {
+	_, file, line := getCallerInfo(skip + 1)
+	if new := l.updateCounter(file, line); new {
+		l.l.Fatal(msg, fields...)
+	}
+}
+
+func Fatal(msg string, fields ...Field) {
+	fatal(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) Fatal(msg string, fields ...Field) {
+	fatal(1, l, msg, fields...)
+}
+
+// Sync
+func (l *Logger) Sync() error {
+	return l.l.Sync()
+}
+
+// Package level default Logger
+var defaultLogger = NewLogger(NewDefaultLoggerConfig())
+
+func Default() *Logger {
+	return defaultLogger
+}
+
+// SetDefault sets package level logger.
+// It's not thread safe so if required use it always at the beggining.
+// Log count of the current default logger will be lost.
+func SetDefault(l *Logger) {
+	if l == nil {
+		panic("Logger cannot be nil")
+	}
+
+	defaultLogger = l
+}
+
+// Init initializes the package level default logger using given config.
+// It's not thread safe so if required use it always at the beggining.
+// Log count of the current default logger will be lost.
+func Init(cfg *LoggerConfig) {
```

Comment:
So I checked for the case where an Init call is missing and messages were simply gone from output. While this is better than panicking, it's important that at the very least it's documented that an Init call must be made, otherwise consuming apps may miss it and miss output logs.
Maybe even an `init` method here would be appropriate, and then `Init` can override the default set logger.

Reply:
Actually if Init() call is missing the logs won't be gone, since `defaultLogger` is [initialized](https://github.com/aquasecurity/tracee/pull/2110/files#diff-c1dc2eb1e149a1dc7ecd5cb474e167fb456eb7ad0ea0256cf62fba693a89af79R243) first. Perhaps, to avoid confusion, Init() should be named `Reset()` and SetDefault() made private `setDefault()`. Does it make sense?

Regarding init(), I agree with you.

Something like:

```go
var defaultLogger *Logger

func init() {
	NewLogger(NewDefaultLoggerConfig())
}
```

Reply:
Strange, since logs were gone when I checked on my machine, but maybe I did something wrong.
Anyways let's go with your `init`.

Reply:
I'll be pushing changes soon. :+1: 

> Strange, since logs were gone when I checked on my machine, but maybe I did something wrong.

Could it be related to avoidance of output when a origin had already been triggered? Currently this approach only writes to output when the log origin is [new](https://github.com/aquasecurity/tracee/pull/2110/files#diff-c1dc2eb1e149a1dc7ecd5cb474e167fb456eb7ad0ea0256cf62fba693a89af79R192-R194):

```go
	_, file, line := getCallerInfo(skip + 1)
	if new := l.updateCounter(file, line); new {
		l.l.Debug(msg, fields...)
	}
```

If the origin has already been triggered, it just updates the counter.